Create a library module for expenses

larmie56 opened the pull request:
What does this PR do?
- Create a library module for expenses and an Expense object

Description of Task to be completed?
- Create a library module for expenses
- Create a domain package in the expense library module
- Create a package for models in the domain package
- Create an Expense object

larmie56 committed:
Create a library module for expenses

Add a domain package to the expense library module

Create an Expense object in the domain package

Ezike reviewed (changes requested):
Nice work. You need to define interfaces/contracts for how other parts of the project will communicate with the domain

larmie56 commented:
> Nice work. You need to define interfaces/contracts for how other parts of the project will communicate with the domain

I will get to that now. Thanks alot boss

larmie56 committed:
Add contract

Define contract for communicating with the domain

Ezike commented:
> > Nice work. You need to define interfaces/contracts for how other parts of the project will communicate with the domain
> 
> 
> 
> I will get to that now. Thanks alot boss

Cool. Now you need to add usecases for the different operations. You can refer to the articles I sent you if you need guidance. Let me know if you have any questions.

Ezike commented on lib_expense/src/main/java/com/example/expenselogger/lib_expense/domain/contract/ExpenseContract.kt:
```diff
@@ -0,0 +1,10 @@
+
+internal interface ExpenseContract {
+    fun saveExpenseEntry(expense: Expense)
+    fun getExpenseEntry(): Expense?
```
When getting an expense item, You need to identify it somehow. Consider adding a unique id property to the expense object. Then you can use the id in this function as a parameter

larmie56 replied:
Yeaaa... skipped my mind. Thank youu

Ezike commented on lib_expense/src/main/java/com/example/expenselogger/lib_expense/domain/contract/ExpenseContract.kt:
```diff
@@ -0,0 +1,10 @@
+import com.example.expenselogger.lib_expense.domain.model.Expense
+
+internal interface ExpenseContract {
+    fun saveExpenseEntry(expense: Expense)
```
Is there a reason you suffixed all the functions with `Entry`? If not, then I think it can be removed since the function names will still be descriptive without it.

larmie56 replied:
I'll remove it from the functions

larmie56 committed:
Modify Expense object

Modify expense contract

larmie56 committed:
Add usecases

Ezike commented on lib_expense/src/main/java/com/example/expenselogger/lib_expense/domain/model/Expense.kt:
```diff
@@ -0,0 +1,11 @@
+    val name: String,
+    val amount: Double,
+    val date: Date,
+    val info: String = ""
```
Why does this have a default value?

larmie56 replied:
Based on the UI, putting info about an expense entry is optional. That's why I gave it a default value, so expense entries without info can be saved, without passing in a value for info

Ezike replied:
Ok that's fine. What we can do is have a factory function for expenses without an information field and avoid the default value in a domain object. 
```
companion object {
   fun create(
    name: String,
    amount: Double,
    date: Date
   ) = Expense(name, amount, date, info=“”)
}
```

Ezike replied:
On a second thought, we can forget bout all of this and just pass an empty string if the info was not provided

larmie56 replied:
Okay boss

Ezike commented on lib_expense/src/main/java/com/example/expenselogger/lib_expense/domain/usecase/EditExpense.kt:
```diff
@@ -0,0 +1,24 @@
+     * @param performEdit - function to perform the edit. It takes the expense to edit as a
+     * parameter, and returns the expense when the edit is done
+     */
+    suspend fun editExpense(id: Long, performEdit: (Expense) -> Expense) {
```
Why did you choose to implement the uscease this way? Afaik when editing an expense, you already have access to the full expense object. This removes the need for a database lookup. The save expense usecase should suffice while the Db handles the update.

larmie56 replied:
Oh okay, I was thinking there won't be access to the full expense object, just the id; and then there'll be a look up to get the expense object (I was considering the argument to the next screen, being as lightweight as possible)
I'll refactor to use the save expense usecase

Ezike replied:
Even if you pass just the id to the edit screen, you first have to get the associated expense object from the Db and render on the screen. So when saving the updated expense, a new expense object can be created from whatever the user entered

larmie56 committed:
Add fetch expense usecase

Refactor expense object

Add save expense usecase

larmie56 committed:
Refactor fetch expense usecase

Ezike committed:
- add base use cases
- enable strict api mode for android libs
- add executor module that wraps coroutine dispatchers

Ezike committed:
refactor
